python/pc_ble_driver_py/ble_adapter.py

```diff
@@ -367,33 +376,60 @@ def write_cmd(self, conn_handle, uuid, data):
-        sec_params  = BLEGapSecParams(bond          = False,
-                                      mitm          = False,
-                                      lesc          = False,
-                                      keypress      = False,
-                                      io_caps       = BLEGapIOCaps.none,
-                                      oob           = False,
-                                      min_key_size  = 7,
-                                      max_key_size  = 16,
+    def authenticate(self, conn_handle, role=BLEGapRoles.central, bond=False, mitm=False, lesc=False,
+                     keypress=False, io_caps=BLEGapIOCaps.none, oob=False, min_key_size=7,
+                     max_key_size=16, enc_own=True, id_own=False, sign_own=False, link_own=False,
+                     enc_peer=True, id_peer=False, sign_peer=False, link_peer=False):
+        kdist_own   = BLEGapSecKDist(enc  = enc_own,
+                                     id   = id_own,
+                                     sign = sign_own,
+                                     link = link_own)
+        kdist_peer  = BLEGapSecKDist(enc  = enc_peer,
+                                     id   = id_peer,
+                                     sign = sign_peer,
+                                     link = link_peer)
+        sec_params  = BLEGapSecParams(bond          = bond,
+                                      mitm          = mitm,
+                                      lesc          = lesc,
+                                      keypress      = keypress,
+                                      io_caps       = io_caps,
+                                      oob           = oob,
+                                      min_key_size  = min_key_size,
+                                      max_key_size  = max_key_size,
                                       kdist_own     = kdist_own,
                                       kdist_peer    = kdist_peer)
 
         self.driver.ble_gap_authenticate(conn_handle, sec_params)
         self.evt_sync[conn_handle].wait(evt = BLEEvtID.gap_evt_sec_params_request)
 
-        self.driver.ble_gap_sec_params_reply(conn_handle, BLEGapSecStatus.success, None, None, None)
+        # sd_ble_gap_sec_params_reply ... In the central role, sec_params must be set to NULL,
+        # as the parameters have already been provided during a previous call to
+        # sd_ble_gap_authenticate.
+        sec_params = None if self.role == BLEGapRoles.central else sec_params
+        self.driver.ble_gap_sec_params_reply(conn_handle,
+                                             BLEGapSecStatus.success,
+                                             sec_params=sec_params)
         result = self.evt_sync[conn_handle].wait(evt = BLEEvtID.gap_evt_auth_status)
+        # If success then keys are stored in self.driver._keyset.
+        if result['auth_status'] ==  BLEGapSecStatus.success:
+            self._keyset = BLEGapSecKeyset.from_c(self.driver._keyset)
```

Comment:
> This would only allow storing one keyset? 
Yes. And, it does not fullfill a general BLE scenario. 
I will try to elaborate on this point:
  One keyset is particular to two devices: 
 - a Central and a Peripheral, according to GAP;
 - an initiator and a responder, according to the Security Manager; or,
 - a master and a slave, according to the link layer.

  On the other hand, an adapter is partly a container of connections (DbConnection objects) and may take the role of Central or Peripheral. 
 In the current implementation, an adapter does not store the peer address of the device that it connects to; that means, it is not able to differentiate between peer devices that establish a connection.

My proposal is as follows: 
- store peer address and role info as part of a DbConnection.
```
class Connection(DbConnection):
    def __init__(self, peer_addr, role):
      super(Connection, self).__init__()
      self.role      = role
      self.peer_addr = peer_add
      self._keyset   = None
```
and
```
    def on_gap_evt_connected(self, ble_driver, conn_handle, peer_addr, role, conn_params):
        self.db_conns[conn_handle]  = Connection(peer_addr, role)
        self.evt_sync[conn_handle]  = EvtSync(events = BLEEvtID)
        self.conn_in_progress       = False

```
Now, we are able to bind keysets with peer devices (as part of authenticate method):
```
        if result['auth_status'] ==  BLEGapSecStatus.success:
            self.db_conns[conn_handle]._keyset = BLEGapSecKeyset.from_c(self.driver._keyset)
```
> A subsequent connection authentication would overwrite these parameters it seems.
 A subsequent connection authentication *with one and the same peer device* must overwrite the keyset. Do you agree?

Reply:
If you find my proposals are Ok and agree on the new functionality then I will do a commit.

Reply:
I agree that a subsequent authentication with the same peer device must overwrite the keyset. There is one corner case though, which is privacy, and in which case the address would require address resolving to correctly identify the peer device. However, privacy could be considered out of scope of this PR.

Reply:
@Darrew I find your proposals ok, you can go ahead with the commits. Thanks for contributing.

Reply:
@bihanssen Commits are done.

Reply:
@Darrew Do you have a gist/snippet that demonstrates usage of the new apis? That would help speed up the PR. Thanks.

Reply:
@bihanssen have a look at [https://gist.github.com/Darrew/b913cc1ee19a579238e83f26075a6625](url)
 * Class BLECentral shall act as a GAP Central Device. 
 * Security procedures are implemented as methods authenticate and encrypt. 
 * Bond information (keys) are stored in a file (KEYSFILE) and retrieve as needed.
Although, the ble_central code needs some refactoring, it works for me.

Reply:
@Darrew thanks, we'll check it out.